### DISCUSSION TITLE
Implementing MinSR algorithm

### DISCUSSION THREAD

**Discussion by chrisrothUT** [netket/netket#1430] [Ideas]
[This paper](https://arxiv.org/pdf/2302.01941.pdf) was very recently posted on Arxiv which seems like a major breakthrough for stochastic reconfiguration (SR) for NQS. The main idea is that the stochastic estimate of the Quantum Fisher Information can be described by a matrix [n_samples, n_samples] instead of [n_params, n_params]. This allows you to do exact SR with matrix inversion so long as you can fit [n_samples, n_samples] into memory. 

I'm happy to lead the charge in implementing this in NetKet. Right now I foresee a technical problem in computing (O O^+) where we cannot fit O into memory.  Ideally we'd like to write something like:

```
def sample_fisher():
  sample_fisher = 0
  for layer in range(final_layer, first_layer):
    grad_psi_hid = backpropogate(grad_psi_hid)
    grad_psi_W = grad(grad_psi_hid)
    sample_fisher = sample_fisher+ matmul(grad_psi_W,g grad_psi_W.conj.T)
```

Does anyone have any idea how to do this kind of thing in Jax?

**Comment by PhilipVinc** [comment]
I already implemented this more than a year ago... 

You want to implement the neural tangent kernel (what they call T) and then you can pull off a few tricks. 

I'll try to find my code...

**Reply by chrisrothUT** [reply]
Yes, it looks like there's been some significant work at google to efficiently compute NTKs https://github.com/google/neural-tangents. 

Actually, It looks like if you just JIT-compile the whole NTK function it won't instantiate the whole Jacobian in memory.

**Reply by PhilipVinc** [reply]
I think that the library you linked is for the infinite width limit (where you don't use a jacobian because... well, it would be infinitely large, but can compute its action under some approximations).

The NTK is literally the gram transpose of the QGT. 
You could implement a Jacobian-based version by copy-pasting the QGTJacobianXXX code and simply transposing the two jacobian vector multiplications

**Reply by chrisrothUT** [reply]
Yeah the NTK function is very straightforward to implement. But I want to make sure the version plays nicely with the GPUs. 

As for the algorithm I'll just put it in `nk.experimental` for the moment and we can think about how it fits into the main library. The issue is that it's preconditioning over the sample dimension so it doesn't fit cleanly into the current interface.

**Reply by PhilipVinc** [reply]
Ah, here I found it. I had hidden it away in a branch named `furbizia` about an year ago 
https://github.com/netket/netket/tree/pv/furbizia

It's similar to what you started prototyping already.
I think the chunking of the NTK should be somewhat equivalent to what `NTKOnTheFly` with default chunking.

**Reply by chrisrothUT** [reply]
Thanks for the link. I actually don't see an `NTKOnTheFly` on that branch but it looks like this is an on the fly implementation where the jacobian or NTK is never fully instantiated unless you call `_to_dense`. Or am I mistaken? This is probably the complementary approach that I haven't coded so it's useful.

**Reply by PhilipVinc** [reply]
No I never implemented it. 
Sorry, I was probably unclear.
What I meant is that `NTKOnTheFly` with chunking would behave as your double loop of computing one sub-part of jacobian. 
Or the opposite. I think that your implementation is somewhat equivalent to what `NTKOnTheFly+chunking` would do.
